### PR TITLE
#3006 image uploader enhancements

### PR DESCRIPTION
**Pull request by AlexeyTuboltsev**
this PR 
- removes delete button 
- add type/dimensions checking before upload

### CODE REVIEW

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 1]
```diff
@@ -31,24 +43,78 @@ export class FormSingleFileUploadComponent extends React.Component<
...
+            reject(createMessage(ImageUploadMessage.IMAGE_UPLOAD_WRONG_IMAGE_TYPE));
+        }),
+      ]);
+
+    const result = await onImageLoad(image);
+    URL.revokeObjectURL(image.src);
+    return result;
+  };
+
+  private checkImage = async (file: File): Promise<TMessage | null> => {
```
Comment (desfero) [inline app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: maybe `validateImage`? but that's up to you

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 2]
```diff
@@ -31,24 +43,78 @@ export class FormSingleFileUploadComponent extends React.Component<
...
+                height: ((e.target as unknown) as IImageDimensions).height,
+              });
+            } else {
+              return reject(createMessage(ImageUploadMessage.IMAGE_UPLOAD_WRONG_IMAGE_TYPE));
+            }
+          };
+        }),
+        new Promise((_, reject) => {
+          image.onerror = () =>
+            reject(createMessage(ImageUploadMessage.IMAGE_UPLOAD_WRONG_IMAGE_TYPE));
```
Comment (desfero) [inline app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: reject should return a custom error and later in `checkImage` in catch we should check for error instance or type. By this, we can make this function more generic

File: app/components/translatedMessages/messages.tsx [hunk 3]
```diff
@@ -158,11 +159,16 @@ export enum IpfsMessage {
...
+}
+
+export enum ImageUploadMessage {
+  IMAGE_UPLOAD_WRONG_IMAGE_TYPE = "imageUploadWrongImageType",
+  IMAGE_UPLOAD_WRONG_IMAGE_DIMENSIONS = "imageUploadWrongImageDimensions",
 }
 
 export enum RemoteFileMessage {
-  GET_FILES_DETAILS_ERROR = "GET_FILES_DETAILS_ERROR",
+  GET_FILES_DETAILS_ERROR = "getFilesDetailsError",
```
Comment (desfero) [inline app/components/translatedMessages/messages.tsx]: 🎖

File: app/modules/notificationModal/sagas.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import { TGlobalDependencies } from "../../di/setupBindings";
+import { actions, TAction } from "../actions";
+import { neuTakeEvery } from "../sagasUtils";
+import { ENotificationModalType } from "./actions";
+
+export function* showNotification(
+  { notificationCenter, logger }: TGlobalDependencies,
+  action: TAction,
+): any {
```
Comment (desfero) [inline app/modules/notificationModal/sagas.ts]: `Iterator<any>`

File: app/modules/notificationModal/sagas.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { fork } from "redux-saga/effects";
+
+import { TGlobalDependencies } from "../../di/setupBindings";
+import { actions, TAction } from "../actions";
+import { neuTakeEvery } from "../sagasUtils";
+import { ENotificationModalType } from "./actions";
+
+export function* showNotification(
+  { notificationCenter, logger }: TGlobalDependencies,
+  action: TAction,
```
Comment (desfero) [inline app/modules/notificationModal/sagas.ts]: `TActionFromFactory<typeof actions.notificationModal.notify>` and then we can remvoe `if` on line `12`

File: app/modules/notificationModal/sagas.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+  const { type, message } = action.payload;
+
+  switch (type) {
+    case ENotificationModalType.ERROR:
+      return notificationCenter.error(message);
+    case ENotificationModalType.INFO:
+      return notificationCenter.info(message);
+    default:
+      logger.error("unknown notification type");
```
Comment (desfero) [inline app/modules/notificationModal/sagas.ts]: maybe `assertNever` to catch this on build time?

File: app/modules/notificationModal/sagas.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  }
+}
+
+export function* notificationModalSagas(): any {
```
Comment (desfero) [inline app/modules/notificationModal/sagas.ts]: `Iterator<any>`

File: app/test-e2e/utils/forms.ts [hunk 8]
```diff
@@ -264,7 +264,8 @@ export const uploadDocumentToFieldWithTid = (targetTid: string, fixture: string)
 export const uploadSingleFileToFieldWithTid = (targetTid: string, fixture: string) => {
   cy.get(tid(targetTid)).within(() => {
     cy.root().dropFile(fixture);
-    cy.get(tid("single-file-upload-delete-file")).should("exist");
+    cy.wait(500);
```
Comment (desfero) [inline app/test-e2e/utils/forms.ts]: What's the reason for `wait` here? `cy.get("img").should` will check in loop until image is there

File: app/components/eto/registration/pages/CompanyInformation.tsx [hunk 9]
```diff
@@ -89,29 +89,32 @@ const EtoRegistrationCompanyInformationComponent = (
...
           />
         </Col>
         <Col>
           <FormSingleFileUpload
+            dimensions={{ width: 1250, height: 400 }}
             name="companyBanner"
             label={<FormattedMessage id="eto.form.company-information.banner" />}
-            acceptedFiles="image/*"
+            acceptedFiles={["image/png", "image/jpeg"]}
             fileFormatInformation="*1250 x 400 png"
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/CompanyInformation.tsx]: wow good idea

File: app/components/eto/registration/pages/CompanyInformation.tsx [hunk 10]
```diff
@@ -89,29 +89,32 @@ const EtoRegistrationCompanyInformationComponent = (
       <Row>
         <Col>
           <FormSingleFileUpload
+            dimensions={{ width: 150, height: 150 }}
             name="companyLogo"
             label={<FormattedMessage id="eto.form.company-information.logo" />}
-            acceptedFiles="image/*"
+            acceptedFiles={["image/png", "image/jpeg"]}
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/CompanyInformation.tsx]: no it's very reasonable, i'll do it

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 11]
```diff
@@ -31,24 +43,78 @@ export class FormSingleFileUploadComponent extends React.Component<
     isUploading: false,
   };
 
-  private onDropFile = (file: File) => {
-    this.setState({ isUploading: true });
+  private getImageDimensions = async (file: File): Promise<IImageDimensions> => {
+    const image = new Image();
+    image.src = URL.createObjectURL(file);
+
+    const onImageLoad = (image: HTMLImageElement): Promise<IImageDimensions> =>
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: 👍

File: app/components/eto/registration/pages/CompanyInformation.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/CompanyInformation.tsx]: done

File: app/components/eto/registration/pages/CompanyInformation.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/CompanyInformation.tsx]: done

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 12]
```diff
@@ -31,24 +43,78 @@ export class FormSingleFileUploadComponent extends React.Component<
     isUploading: false,
   };
 
-  private onDropFile = (file: File) => {
-    this.setState({ isUploading: true });
+  private getImageDimensions = async (file: File): Promise<IImageDimensions> => {
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: done

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: done

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 13]
```diff
@@ -31,24 +43,78 @@ export class FormSingleFileUploadComponent extends React.Component<
...
+    const image = new Image();
+    image.src = URL.createObjectURL(file);
+
+    const onImageLoad = (image: HTMLImageElement): Promise<IImageDimensions> =>
+      Promise.race([
+        new Promise((resolve, reject) => {
+          image.onload = (e: Event) => {
+            if (e.target !== null) {
+              return resolve({
+                width: ((e.target as unknown) as IImageDimensions).width,
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: done

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx (hunk 2 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: done

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx (hunk 1 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: yes, it's a good point

File: app/modules/notificationModal/sagas.ts (hunk 5 above)
Comment (AlexeyTuboltsev) [reply app/modules/notificationModal/sagas.ts]: do you mean `TActionFromCreator`? We should rename it, it sounds like smth religious 😁

File: app/modules/notificationModal/sagas.ts (hunk 4 above)
Comment (AlexeyTuboltsev) [reply app/modules/notificationModal/sagas.ts]: done

File: app/modules/notificationModal/sagas.ts (hunk 6 above)
Comment (AlexeyTuboltsev) [reply app/modules/notificationModal/sagas.ts]: done

File: app/modules/notificationModal/sagas.ts (hunk 7 above)
Comment (AlexeyTuboltsev) [reply app/modules/notificationModal/sagas.ts]: done

File: app/test-e2e/utils/forms.ts (hunk 8 above)
Comment (AlexeyTuboltsev) [reply app/test-e2e/utils/forms.ts]: removed

File: app/components/shared/SingleFileUpload.tsx [hunk 14]
```diff
@@ -66,30 +61,21 @@ export class SingleFileUpload extends React.Component<IProps & CommonHtmlProps>
...
         <div className={styles.fakeDropzoneArea}>
-          {hasFile ? <img src={file} alt="File uploaded" /> : dropzoneInner}
+          {file !== undefined ? <img src={file} alt="File uploaded" /> : dropzoneInner}
         </div>
         <div className={styles.sideBox}>
-          {!disabled &&
-            (hasFile ? (
+          {!disabled && (
+            <>
               <Button
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/SingleFileUpload.tsx]: oh then I have to keep `onDeleteFile` and all types and the delete logic inside `FormSingleFileUPload` too. This is a huge chunk of code that most likely will never be needed or will be needed in some indefinite future when it becomes outdated and will have to be refactored anyway. I'd keep our codebase clean, I'd say it's more important. What do you think?

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 15]
```diff
@@ -5,17 +5,22 @@ import { compose } from "recompose";
...
+  TAcceptedFileType,
+} from "./utils.unsafe";
 
 interface IOwnProps {
   disabled?: boolean;
   name: string;
   label: string | React.ReactNode;
-  acceptedFiles: TAcceptedFileType | TAcceptedFileType[];
-  fileFormatInformation: string;
+  acceptedFiles: [TAcceptedFileType, ...TAcceptedFileType[]]; //at least one required
```
[the remark this answers is not in the document]
Comment (desfero) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: Same should probably go to `SingleFileUpload`?

File: app/modules/notificationModal/sagas.ts (hunk 5 above)
Comment (desfero) [reply app/modules/notificationModal/sagas.ts]: yeah, agree, will rename it from `TFromAction` after i'm finishing with the refactoring of old `createAction`s

File: app/components/shared/SingleFileUpload.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (desfero) [reply app/components/shared/SingleFileUpload.tsx]: yeah, then let's do that way, i'm not very opinionated here

File: app/components/shared/forms/fields/utils.unsafe.tsx [hunk 16]
```diff
@@ -91,3 +103,82 @@ export const generateMaskFromCurrency = (currency: ECurrency, isPrice?: boolean)
...
+      return "";
+    } else if (index !== 0 && index === length - 1) {
+      return " or ";
+    } else {
+      return ", ";
+    }
+  };
+
+  if (fileFormatInformation) {
+    return fileFormatInformation;
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/utils.unsafe.tsx]: done

File: app/components/shared/forms/fields/utils.unsafe.tsx [hunk 17]
```diff
@@ -91,3 +103,82 @@ export const generateMaskFromCurrency = (currency: ECurrency, isPrice?: boolean)
...
+  } else {
+    const imageDimensions = dimensions ? `${dimensions.width}px × ${dimensions.height}px, ` : "";
+
+    const fileTypes = acceptedFiles.reduce(
+      (acc: string, mimeType: string, index: number): string => {
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/utils.unsafe.tsx]: done

File: app/components/shared/forms/fields/utils.unsafe.tsx [hunk 18]
```diff
@@ -91,3 +103,82 @@ export const generateMaskFromCurrency = (currency: ECurrency, isPrice?: boolean)
...
+    };
+    image.onerror = () => reject(new Error("image upload failed"));
+    image.onabort = () => reject(new Error("image upload failed"));
+  });
+
+export const readImageAndGetDimensions = async (file: File): Promise<IImageDimensions> => {
+  const image = new Image();
+  image.src = URL.createObjectURL(file);
+
+  return await onImageLoad(image).then(res => {
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/utils.unsafe.tsx]: you're right

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx [hunk 19]
```diff
@@ -43,43 +48,16 @@ export class FormSingleFileUploadComponent extends React.Component<
...
-  private checkImage = async (file: File): Promise<TMessage | null> => {
+  private validateImage = async (file: File): Promise<TMessage | null> => {
     //null for no error -- go style :)
     if (this.props.dimensions) {
       try {
-        const dimensions = await this.getImageDimensions(file);
-        if (
-          dimensions.width === this.props.dimensions.width &&
-          dimensions.height === this.props.dimensions.height
+        const readImageResult: IImageDimensions | Error = await readImageAndGetDimensions(file);
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: you're right

File: app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/FormSingleFileUpload.unsafe.tsx]: done done

File: app/components/shared/forms/fields/utils.unsafe.tsx [hunk 20]
```diff
@@ -91,3 +103,82 @@ export const generateMaskFromCurrency = (currency: ECurrency, isPrice?: boolean)
...
+        return acc;
+      },
+      "",
+    );
+
+    return `${imageDimensions}${fileTypes}`;
+  }
+};
+
+//this one won't have a unit tests because JSDOM doesn't support image.onload and URL.createObjectURL
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/fields/utils.unsafe.tsx]: left a todo

**Pull request merged**